wireapp/wire-ios
# fix: can't login after switching backend via deep link FS-674

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/FS-674" title="FS-674" target="_blank"><img alt="Bug" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10803?size=medium" />FS-674</a>  [iOS] Endless spinner when logging in to Federated Backend through Deeplink
  </td></table>
  <br />
 






----
#### PR Submission Checklist for internal contributors

- The **PR Title**

- The **PR Description**
----

# What's new in this PR?

### Issues

When switching to a custom backend with a deep link, I login in with an existing account on that backend, the load spinner appears forever and the flow is stuck.

### Causes

Actually the app successfully logs the user in.  The `SessionManager` will then try to inform it's `loginDelegate` (an instance of `AuthenticationCoordinator` so that it can proceed to the next step in the authentication flow. However, the `loginDelegate` is `nil`, so the flow never progresses.

The problem is:
- When the app is launched the first time, there exists an instance of `AuthenticationCoordinator` (instance A) which is the `loginDelegate`
- When processing the deep link, a new instance (B) of `AuthenticationCoordinator` is created and replaces A. B is now the `loginDelegate`.
- At the same time, after switch backends, the `SessionManager` creates a new unauthenticated session, which triggers some observer code in instance A to set itself as the `loginDelegate` again.
- At this point, `loginDelegate` is A.
- But because B replaces A, A is released from memory, and `loginDelegate` is set to `nil`.

### Solutions

Before replacing the authentication coordinator, or setting it to `nil`, tear it down by discarding all of its observers. This ensures that no observer code is called after we've created a new coordinator.

### Testing

#### Test Coverage

- We tried to add test coverage, but `AuthenticationCoordinator` is extremely hard to test due to its large amount of dependencies.
- We want to address this however.

#### How to Test

- Install the app fresh
- Switch to new backend via deep link
- Log in to new backend
- Verify that you reach the conversation list.

----
#### PR Post Submission Checklist for internal contributors (Optional)

----
#### PR Post Merge Checklist for internal contributors

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.
